File: WaterArea.java
// Time Complexity : n^2
//Space Complexity : O(1) 
//Did this code successfully run on Leetcode : Yes

class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        int area = Integer.MIN_VALUE;
        for(int i = 0; i < n -1 ; i ++ ){
            for (int j = n - 1; j > i; j--){
                int width = j - i;
                int h = Math.min(height[i], height[j]);
                int temp = width * h;
                if(temp > area){
                    area = temp;
                }
            }
        }
        return area;
    }
}
// Time Complexity : O(n)
//Space Complexity : O(1) 
//Did this code successfully run on Leetcode : Yes
//-------------Using Two pointer High and low, if low is equal to minimun increment low else increment high, As shown in class-------------------
class Solution {
    public int maxArea(int[] height) {
        int n = height.length;
        int area = Integer.MIN_VALUE;
        int min = Integer.MIN_VALUE;
        int low = 0;
        int high = n -1;
        while(low < high){
            int minH = Math.min(height[low], height[high]);
            int temp = (high - low) * minH;
            if(minH == height[low]){
                low++;
            }
            else{
                high--;
            }
            
            if(area < temp){
                area = temp;
            }
            
        }
        return area;
    }
}

File: waterArea.py
"""
Intuition: Area of a rectangle = width * height
Here width is the difference in the index. and height is the minimun of 2 heights. 
We pick the minimum of 2 heights as otherwise water will overflow.

In 2 pointer approach, the width will always decrease. So we need to maximize of height.
Hence we alawys move the pointer where there is a probability of increasing the height.
#####################################################################
Time Complexity : O(N)  N = number of elements
Space Complexity : O(1) 
#####################################################################
"""
class Solution:
    def maxArea(self, heights: List[int]) -> int:
        
        maxArea = 0
        left = 0
        right = len(heights) - 1
        
        while left < right:
            
            area = min(heights[left], heights[right])* (right - left)
            maxArea = max(area, maxArea)
            
            if heights[left] < heights[right]:
                left += 1
            else:
                right -= 1
            
        return maxArea